apps/condo/domains/resident/tasks/helpers/sendResidentMessage.js

```diff
@@ -0,0 +1,181 @@
+ * @param data
+ * @returns {*}
+ */
+const renderTemplateString = (templateString, data) => {
+    const keys = Object.keys(data)
+    let result = templateString
+
+    for (const urlKey of keys) {
+        const value = get(data, urlKey)
+
+        if (value) result = result.replace(`{${urlKey}}`, value)
+    }
+
+    return result
+}
+
+/**
+ * Fills notification payload from request data according to message meta settings by selected notification type
+ * @param messageType
+ * @param notificationData
+ * @param data
+ */
+const fillExtraData = (messageType, notificationData) => {
+    const metaKeys = Object.keys(get(MESSAGE_META[messageType], 'data', {}))
+    const result = {}
+
+    for (const fieldKey of metaKeys) {
+        const value = get(notificationData, fieldKey)
+
+        if (value) result[fieldKey] = value
+    }
+
+    return result
+}
+
+/**
+ * Please have a look at SendResidentMessageService for json structure
+ * @param json
+ * @returns {Promise<{attemptsCount: (*), successCount: number}|{attemptsCount: number, successCount: number, json, error: Error}>}
+ */
+const sendResidentMessage = async (json) => {
+    let data
+
+    try {
+        data = JSON.parse(json)
+    } catch (error) {
+        logger.error({ msg: 'Error parsing JSON.', error: error.message })
+
+        return { attemptsCount: 0, successCount: 0, error, json }
+    }
+
+    const { organizationId, propertyDetails, uniqKeyTemplate, type: messageType, data: notificationData = {} } = data
+    const { title, message, urlTemplate, categoryId } = notificationData
+    const residentsWhere = { organization: { id: organizationId }, deletedAt: null }
+    const { keystone: context } = await getSchemaCtx('BillingProperty')
+    let category = {}
+
+    if (categoryId) {
+        category = await BillingCategory.getOne(context, { id: categoryId })
+
+        if (isEmpty(category)) {
+            logger.error({ msg: 'Category id is unknown', categoryId })
+
+            return { attemptsCount: 0, successCount: 0, error: new Error(`Category id is unknown: ${categoryId}`), json }
+        }
+    }
+
+    if (isArray(propertyDetails)) {
+        const billingPropertyIds = uniq(compact(propertyDetails.map(({ billingPropertyId }) => billingPropertyId)))
+        let propertiesFromBilling = []
+
+        // details have some billingProperties, so we should convert billingPropertyIds to propertyIds
+        if (!isEmpty(billingPropertyIds)) {
+            const billingPropertyWhere = {
+                id_in: billingPropertyIds,
+                deletedAt: null,
+            }
+            const billingProperties = await loadListByChunks({ context, list: BillingProperty, where: billingPropertyWhere })
+
+            if (!isEmpty(billingProperties)) {
+                const addresses = uniq(compact(billingProperties.map(({ normalizedAddress }) => normalizedAddress)))
+                const addresses1 = uniq(compact(billingProperties.map(({ address }) => address)))
```

Comment:
What is `addresses1`?
```suggestion
                const normalizedAddresses = uniq(compact(billingProperties.map(({ normalizedAddress }) => normalizedAddress)))
                const addresses = uniq(compact(billingProperties.map(({ address }) => address)))
```